analytical_engine/apps/assortativity/degree/degree_assortativity_coefficient.h

```diff
@@ -0,0 +1,208 @@
+/** Copyright 2020 Alibaba Group Holding Limited.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef ANALYTICAL_ENGINE_APPS_ASSORTATIVITY_DEGREE_DEGREE_ASSORTATIVITY_COEFFICIENT_H_
+#define ANALYTICAL_ENGINE_APPS_ASSORTATIVITY_DEGREE_DEGREE_ASSORTATIVITY_COEFFICIENT_H_
+
+#include <unordered_map>
+#include <vector>
+#include <numeric>
+
+#include "apps/assortativity/degree/degree_assortativity_coefficient_context.h"
+#include "core/app/app_base.h"
+#include "core/worker/default_worker.h"
+#include "grape/grape.h"
+
+namespace gs {
+/**
+ * @brief Compute the degree assortativity for graph.
+ *  Assortativity measures the similarity of connections in the graph with
+ * respect to the node degree.
+ * @tparam FRAG_T
+ */
+template <typename FRAG_T>
+class DegreeAssortativity
+    : public AppBase<FRAG_T, DegreeAssortativityContext<FRAG_T>>,
+      public grape::Communicator {
+ public:
+  INSTALL_DEFAULT_WORKER(DegreeAssortativity<FRAG_T>,
+                         DegreeAssortativityContext<FRAG_T>, FRAG_T)
+  static constexpr grape::MessageStrategy message_strategy =
+      grape::MessageStrategy::kAlongEdgeToOuterVertex;
+  static constexpr grape::LoadStrategy load_strategy =
+      grape::LoadStrategy::kBothOutIn;
+  using vertex_t = typename fragment_t::vertex_t;
+  using vid_t = typename fragment_t::vid_t;
+  using oid_t = typename fragment_t::oid_t;
+
+  void PEval(const fragment_t& frag, context_t& ctx,
+             message_manager_t& messages) {
+    ctx.merge_stage = false;
+    // vid
+    auto inner_vertices = frag.InnerVertices();
+    int source_degree, dest_degree;
+    // w of type: Vertex
+    for (auto w : inner_vertices) {
+      source_degree = GetDegreeByType(frag, w, ctx.source_degree_type_);
+      // updata max degree
+      ctx.max_degree =
+          ctx.max_degree > source_degree ? ctx.max_degree : source_degree;
+      // vid_t source_vid = frag.Vertex2Gid(w);
+      VLOG(0) << "vertex " << frag.GetId(w) << " InDegree:" << source_degree
+              << std::endl;
+      // get all neighbors of vertex w
+      auto oes = frag.GetOutgoingAdjList(w);
+      for (auto& e : oes) {
+        vertex_t neighbor = e.get_neighbor();
+        if (frag.IsOuterVertex(neighbor)) {
+          vid_t dest_vid = frag.Vertex2Gid(neighbor);
+          fid_t fid = frag.GetFragId(neighbor);
+          std::pair<vid_t, int> send_msg(dest_vid, source_degree);
+          messages.SendToFragment(fid, send_msg);
```

Comment:
use SendMsgThroughOEdges here, and just send source_degree.

Reply:
if the graph is undirected graph, here u, v would send to each other, and process the degree-degree pair of the edge twice.

Reply:
In fact, for undirected graph, we need process the degree-degree pair of the edge twice, because an undirected edge is equivalent to two edges in directed graph.